### PR TITLE
[FS-1041] Add IReadOnlyCollection/List/Dictionary implementations to FSharp.Core collections

### PR DESCRIPTION
**Pull request by saul**
Implements RFC https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1041-ireadonly.md along with https://github.com/Microsoft/visualfsharp/pull/3988

### CODE REVIEW

**Comment by cartermp** [comment]
https://github.com/fsharp/fslang-design/pull/234

File: src/fsharp/FSharp.Core/math/z.fs [hunk 1]
```diff
@@ -402,49 +402,6 @@ namespace Microsoft.FSharp.Core
                 if ok then 
                     res 
                 else 
-#if FSCORE_PORTABLE_OLD
```
Comment (KevinRansom) [inline src/fsharp/FSharp.Core/math/z.fs]: Nice

**Comment by ErikSchierboom** [comment]
Great addition!

**Comment by saul** [comment]
Is this waiting for something from me before it can be merged?

**Comment by vasily-kirichenko** [comment]
@saul I could tell you what it's waiting for, but it'd not polite enough :)

**Comment by KevinRansom** [comment]
@saul  this is a candidate for F# 4.2.  So it is waiting ... for me to rev the version of FSharp.Core in the build.  Actually the version is going to be 4.3 to align with the nugget package but that is neither here nor there.

The wait is because I have not yet revved FSharp.Core's version number.  That is happening in the dev15.6 branch, and so these 4.2 PR's will be retargeted to dev15.6.

And what was in Vasily's mind is pretty accurate ... I have been putting off doing that work for about a week or so right now.

Kevin

**Comment by KevinRansom** [comment]
Changing target to dev15.6

**Pull request merged**